esp32/mods/esp_espnow.c

```diff
@@ -0,0 +1,359 @@
+    mp_obj_base_t base;
+    uint8_t number;
+} mp_obj_espnow_t;
+
+NORETURN void _esp_espnow_exceptions(esp_err_t e) {
+   switch (e) {
+      case ESP_ERR_ESPNOW_NOT_INIT:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Not Initialized");
+      case ESP_ERR_ESPNOW_ARG:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Invalid Argument");
+      case ESP_ERR_ESPNOW_NO_MEM:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Out Of Mem");
+      case ESP_ERR_ESPNOW_FULL:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Peer List Full");
+      case ESP_ERR_ESPNOW_NOT_FOUND:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Peer Not Found");
+      case ESP_ERR_ESPNOW_INTERNAL:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Internal");
+      case ESP_ERR_ESPNOW_EXIST:
+        mp_raise_msg(&mp_type_OSError, "ESP-Now Peer Exists");
+      default:
+        nlr_raise(mp_obj_new_exception_msg_varg(
+          &mp_type_RuntimeError, "ESP-Now Unknown Error 0x%04x", e
+        ));
+   }
+}
+
+static inline void esp_espnow_exceptions(esp_err_t e) {
+    if (e != ESP_OK) _esp_espnow_exceptions(e);
+}
+
+#define ESPNOW_EXCEPTIONS(x) do { esp_espnow_exceptions(x); } while (0);
+
+static inline void _get_bytes(mp_obj_t str, size_t len, uint8_t *dst) {
+    size_t str_len;
+    const char *data = mp_obj_str_get_data(str, &str_len);  
+    if (str_len != len) mp_raise_ValueError("bad len");
+    memcpy(dst, data, len);
+}
+
+static mp_obj_t send_cb_obj = mp_const_none;
+static mp_obj_t recv_cb_obj = mp_const_none;
+
+
+STATIC void send_queue_handler(void *arg) {
+    // this function will be called by the interrupt thread
+    mp_obj_tuple_t *msg = arg;
+    if (send_cb_obj != mp_const_none) {
+        mp_call_function_1(send_cb_obj, msg);
+    }
+}
+
+STATIC void IRAM_ATTR send_cb(const uint8_t *macaddr, esp_now_send_status_t status)
+{
+    if (send_cb_obj != mp_const_none) {
+        mp_obj_tuple_t *msg = mp_obj_new_tuple(2, NULL);
+        msg->items[0] = mp_obj_new_bytes(macaddr, ESP_NOW_ETH_ALEN);
+        msg->items[1] = (status == ESP_NOW_SEND_SUCCESS) ? mp_const_true : mp_const_false;
+        mp_irq_queue_interrupt(send_queue_handler, msg);
+    }
+}
+
+STATIC void recv_queue_handler(void *arg) {
+    // this function will be called by the interrupt thread
+    mp_obj_tuple_t *msg = arg;
+    if (recv_cb_obj != mp_const_none) {
+        mp_call_function_1(recv_cb_obj, msg);
+    }
+}
+
+STATIC void IRAM_ATTR recv_cb(const uint8_t *macaddr, const uint8_t *data, int len) 
+{
+    if (recv_cb_obj != mp_const_none) {
+        mp_obj_tuple_t *msg = mp_obj_new_tuple(2, NULL);
+        msg->items[0] = mp_obj_new_bytes(macaddr, ESP_NOW_ETH_ALEN);
+        msg->items[1] = mp_obj_new_bytes(data, len);
+        mp_irq_queue_interrupt(recv_queue_handler, msg);
+    }
+} 
+
+STATIC mp_obj_t espnow_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+ 
+    mp_obj_espnow_t *self = m_new_obj(mp_obj_espnow_t);
```

Comment:
I think it would be better if this class would be "singleton", like e.g. CoAp module. As I understand instantiating again this class would result that esp_now is re-initialized which may ruin the already ongoing former instances.

Reply:
OK - I agree - perhaps like this??

```
// A static pointer to the espnow singleton
STATIC mp_obj_espnow_t *espnow_singleton = NULL;

STATIC mp_obj_t espnow_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {

    if (espnow_singleton != NULL) {
        return espnow_singleton;
    } 

    mp_obj_espnow_t *self = m_new_obj(mp_obj_espnow_t);
    self->base.type = type;
    self->number = 0;
    
    ESPNOW_EXCEPTIONS(esp_now_init());  
    ESPNOW_EXCEPTIONS(esp_now_register_recv_cb(recv_cb));
    ESPNOW_EXCEPTIONS(esp_now_register_send_cb(send_cb));

    espnow_singleton = self;

    return MP_OBJ_FROM_PTR(self);
}
```

> l